core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,228 @@
+	ocrtypes "github.com/smartcontractkit/chainlink/offchainreporting/types"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/pkg/errors"
+)
+
+var (
+	OCRContractConfigSet = getConfigSetHash()
+)
+
+type (
+	LogConfigSet struct {
+		eth.GethRawLog
+		ConfigCount  uint64
+		Oracles      []gethCommon.Address
+		Transmitters []gethCommon.Address
+		Threshold    uint8
+		Config       []byte
+	}
+
+	OCRContract struct {
+		ethClient        eth.Client
+		configSetChan    chan offchainaggregator.OffchainAggregatorConfigSet
+		contractFilterer *offchainaggregator.OffchainAggregatorFilterer
+		contractCaller   *offchainaggregator.OffchainAggregatorCaller
+		contractAddress  gethCommon.Address
+		logBroadcaster   eth.LogBroadcaster
+
+		jobID       models.ID
+		transmitter Transmitter
+		contractABI abi.ABI
+	}
+
+	Transmitter interface {
+		CreateEthTransaction(ctx context.Context, toAddress gethCommon.Address, payload []byte) error
+		FromAddress() gethCommon.Address
+	}
+)
+
+var (
+	_ ocrtypes.ContractConfigTracker      = &OCRContract{}
+	_ ocrtypes.ContractTransmitter        = &OCRContract{}
+	_ ocrtypes.ContractConfigSubscription = &OCRContractConfigSubscription{}
+)
+
+func NewOCRContract(address gethCommon.Address, ethClient eth.Client, logBroadcaster eth.LogBroadcaster, jobID models.ID, transmitter Transmitter) (o *OCRContract, err error) {
+	contractFilterer, err := offchainaggregator.NewOffchainAggregatorFilterer(address, ethClient)
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorFilterer")
+	}
+
+	contractCaller, err := offchainaggregator.NewOffchainAggregatorCaller(address, ethClient)
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorCaller")
+	}
+
+	contractABI, err := abi.JSON(strings.NewReader(offchainaggregator.OffchainAggregatorABI))
+	if err != nil {
+		return o, errors.Wrap(err, "could not get contract ABI JSON")
+	}
+
+	return &OCRContract{
+		contractFilterer: contractFilterer,
+		contractCaller:   contractCaller,
+		contractAddress:  address,
+		logBroadcaster:   logBroadcaster,
+		ethClient:        ethClient,
+		configSetChan:    make(chan offchainaggregator.OffchainAggregatorConfigSet),
+		jobID:            jobID,
+		transmitter:      transmitter,
+		contractABI:      contractABI,
+	}, nil
+}
+
+func (ra *OCRContract) Transmit(ctx context.Context, report []byte, rs, ss [][32]byte, vs [32]byte) error {
+	payload, err := ra.contractABI.Pack("transmit", report, rs, ss, vs)
+	if err != nil {
+		return errors.Wrap(err, "abi.Pack failed")
+	}
+
+	return errors.Wrap(ra.transmitter.CreateEthTransaction(ctx, ra.contractAddress, payload), "failed to send Eth transaction")
+}
+
+func (ra *OCRContract) SubscribeToNewConfigs(context.Context) (ocrtypes.ContractConfigSubscription, error) {
+	sub := &OCRContractConfigSubscription{
+		make(chan ocrtypes.ContractConfig),
```

Comment:
Could this block the whole node if the receiver messes up and stops listening on the channel?

Reply:
Yes, I think it could.

Reply:
My personal feeling on this is that since this is handled in core and not handed off to an external libary, any hangs are an engineering fault on our side. We should just engineer it so this doesn't happen.

Reply:
What about a very long timeout for the worse case scenario?

Reply:
Given the potentially grave consequences, I'd lean towards some kind of graceful failure in case of an infinite block, eg using a timeout like jb suggested.

Reply:
+1 for timeout.  How about we just use the context that's passed into this method?

Reply:
The channel can live for the entire lifetime of the Subscription, so I don't think that we can use the context.

Reply:
What is that context actually supposed to control?

Reply:
The subscription creation

Reply:
I added a timeout. Going to ignore the passed context for now.